### PR TITLE
[Feature](bangc-ops):add new op tin_shift forward and backward

### PR DESCRIPTION
**Pull request by defei-coder**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Add new op tin_shift forward and tin_shift backward.

## 2. Modification

Modify files:
- bangc-ops/kernel_depends.toml
- bangc-ops/mlu_op.h

Add new files:
- bangc-ops/kernels/tin_shift/tin_shift.cpp
- bangc-ops/kernels/tin_shift/tin_shift.h
- bangc-ops/kernels/tin_shift/tin_shift_union1.mlu
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/tin_shift/test_case/case0.prototxt
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/tin_shift/tin_shift.cpp
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/tin_shift/tin_shift.h
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/tin_shift_backward/test_case/case0.prototxt
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/tin_shift_backward/tin_shift_backward.cpp
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/tin_shift_backward/tin_shift_backward.h
- docs/bangc-docs/design_docs/tin_shift/demo1.png
- docs/bangc-docs/design_docs/tin_shift/demo2.png
- docs/bangc-docs/design_docs/tin_shift/tin_shift.md

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff3: diff3 == 0

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |      U1       |
|        3       |Layouts                               |    ARRAY    |
|        4       |Whether multi-dimensions are supported|        NO              |
|        5       |Whether element zero is supported     |     input和output的第二维支持0元素，返回MLUOP_STATUS_SUCCESS；其余维度不支持0元素,返回MLUOP_STATUS_BAD_PARAM   |
|        6       |Data type(half/float)                 |           half / float           |
|        7       |Whether there is size limit           |          NO           |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multi-dimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| input的维度必须是4 |     Normal error    |     [MLUOP] [Error]:[mluOpTinShift forward] The input dims should be 4. But now the input dims is 3.<br>  [Error]:[mluOpTinShift backward] The input dims and the output dims should be the same. But now the input dims is 4, and the output dims is 3.  |
| shift的维度必须是2  |     Normal error    |   [Error]:[mluOpTinShift forward] The shifts dims should be 2. But now the shifts dims is 1. <br> [Error]:[mluOpTinShift backward] The shifts dims should be 2. But now the shifts dims is 1. |
|  input和output的维度必须相同   |     Normal error    |   [Error]:[mluOpTinShift forward] The input dims and the output dims should be the same. But now the input dims is 4, and the output dims is 3.<br> [Error]:[mluOpTinShift backward] The input dims and the output dims should be the same. But now the input dims is 4, and the output dims is 3.  |
|  input和output的每一维的维度必须相同   |     Normal error    | [Error]:[mluOpTinShift forward] The size of input dims[1] is 15, and the size of output dims[1] is 1. They should be the same.<br> [Error]:[mluOpTinShift backward] The size of input dims[1] is 15, and the size of output dims[1] is 1. They should be the same.|
|  input的第三维不能为0   |     Normal error    |  [Error]:[mluOpTinShift forward] The channel size should not be zero. <br> [Error]:[mluOpTinShift backward] The channel size should not be zero. |
|  shift的第二维不能为0   |     Normal error    |  [Error]:[mluOpTinShift forward] The group size should not be zero. <br> [Error]:[mluOpTinShift backward] The group size should not be zero. |
|  input的第三维必须是shift的第二维的整数倍   |     Normal error    |  [Error]:[mluOpTinShift forward] The channel size should be multiple of group size, But now channel size is 16 and group size is 7. <br> [Error]:[mluOpTinShift backward] The channel size should be multiple of group size, But now channel size is 16 and group size is 3. |
|  input的第一维必须和shift的第一维相等   |     Normal error    |  [Error]:[mluOpTinShift forward] The input batch size should be the same as shifts's, input batch size is 3 and shifts batch size is 2. <br> [Error]:[mluOpTinShift backward] The input batch size should be the same as shifts's, input batch size is 3 and shifts batch size is 1. |
|  shift仅支持int32   |     Normal error    |  [Error]:[mluOpTinShift forward] The data type of the shift tensor should be MLUOP_DTYPE_INT32. Current data type is DTYPE_HALF <br> [Error]:[mluOpTinShift backward] The data type of the shift tensor should be MLUOP_DTYPE_INT32. Current data type is DTYPE_FLOAT |
| input仅支持half或float |     Normal error    |  [Error]:[mluOpTinShift forward] The data type of the input tensor should be half or float. <br> [Error]:[mluOpTinShift backward] The data type of the input tensor should be half or float.  |
| input和output的数据类型必须一致    |     Normal error    |  [Error]:[mluOpTinShift forward] The data type of the input tensor and output tensor should be the same. <br> [Error]:[mluOpTinShift backward] The data type of the input tensor and output tensor should be the same. |
| input的第一维不能为0    |     Normal error    | [Error]:[mluOpTinShift forward] The input batch size should not be zero. <br> [Error]:[mluOpTinShift backward] The input batch size should not be zero. |
| input的最后一维不能为0    |     Normal error    |   [Error]:[mluOpTinShift forward] The size of input hw size should not be zero.<br>[Error]:[mluOpTinShift backward] The size of input hw size should not be zero.  |
| 输入输出均不支持元素个数超过2G    |     Normal error    |  [Error]:[mluOpTinShift forward]  overflow max supported tensor num 2147483647, now tensor's total num is 3271320000.  <br>  [Error]:[mluOpTinShift backward]  overflow max supported tensor num 2147483647, now tensor's total num is 3271320000. |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float                 |   [ SUMMARY  ] Total 634 cases of 2 op(s).<br>ALL PASSED.<br>[==========] 634 test cases from 2 test suites ran. (261013 ms total)<br>[  PASSED  ] 634 test cases.  |          |
|Mult-tensor test     | Only support 4 dims   |  Only support 4 dims |     [ SUMMARY  ] Total 634 cases of 2 op(s).<br>ALL PASSED.<br>[==========] 634 test cases from 2 test suites ran. (261013 ms total)<br>[  PASSED  ] 634 test cases.       |
|Layout test          | Supports ARRAY  |    [ SUMMARY  ] Total 634 cases of 2 op(s).<br>ALL PASSED.<br>[==========] 634 test cases from 2 test suites ran. (261013 ms total)<br>[  PASSED  ] 634 test cases.      |          |
|Zero element test    |Whether to support this test      |   部分维度支持，结果见3.1  |          |
|Stability test       |--gtest_repeat=10<br>--thread=10|   [----------] Global test environment tear-down<br>[ SUMMARY  ] Total 6340 cases of 650 op(s).<br>ALL PASSED.<br>[==========] 65 test cases from 65 test suites ran. (27083 ms total)<br>[  PASSED  ] 65 test cases.  |          |
|Mult-platform test   |MLU370/MLU590                     |   [ SUMMARY  ] Total 634 cases of 2 op(s).<br>ALL PASSED.<br>[==========] 634 test cases from 2 test suites ran. (402936 ms total)<br>[  PASSED  ] 634 test cases.  |          |
|Nan/INF test         |Whether to support this test      |   YES  |          |
|Memory leak check    |Test result                       |   [ SUMMARY  ] Total 634 cases of 2 op(s).<br>ALL PASSED.<br>[==========] 634 test cases from 2 test suites ran. (402936 ms total)<br>[  PASSED  ] 634 test cases.  |          |
|Code coverage check  |Test result                       |      kernels/tin_shift  [95.9%][95.9%] 95.9 % 233 / 243      100.0 %9 / 9  |   switch default无法覆盖       |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|tin_shift_forward| 450 | 39 | 0.946563 | 0.0354961 | 0 | half | input : [3,15,16,45435] <br> shifts:[3,8]|
|tin_shift_backward| 2443 | 43.95 | 1.00598 | 0.0377241 | 0 | float | input:[3,15,16,262144] <br> shifts:[3,8] |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|tin_shift_forward| 50 | 28.434 | 0.946563 | 0.13311 | 0 | half | input : [3,15,16,45435] <br> shifts:[3,8]|
|tin_shift_backward| 263 | 201.58 | 1.03828 | 0.146008 | 0 | float | input:[3,15,16,262144] <br> shifts:[3,8] |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: docs/bangc-docs/design_docs/tin_shift/tin_shift.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+
+* #### 文档基本信息
+
+| 算子名称  | mluOpTinShiftBackward && mluOpTinShiftForward        |
+| ----------- | ------------------------------------- |
+| 编制人/日期 | 卜德飞/2021-09-27                      |
+| 审批人/日期 | 周晨阳/2021-11-11                      |
+| 审批人/日期 | 鲁德辉/2021-11-11                      |
+| 审批人/日期 | 邓皓煜/2021-11-11                      |
```
Comment (PetrelYy) [inline docs/bangc-docs/design_docs/tin_shift/tin_shift.md:11]: 缺少 user_guide/9_idx.rst 修改

**Review by PetrelYy — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -12863,6 +12863,132 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ *   - shifts tensor: int32
+ *   - output tensor: half, float
+ *
+ * @par Data Layout
+ * - The supported layout of input and output tensors must be \p MLUOP_LAYOUT_ARRAY.
+ *
+ * @par Scale Limitation
+ * - The first dimension of \b grad_output tensor and \b shifts tensor must be the same size.
+ * - The second dimension of \b grad_output tensor must be multiple of the second dimension of \b shifts tensor.
+ *   For example, if the shape of grad_output is [N, T, C, HW], the shape of shifts is [N, G],
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:12905]:
```suggestion
 *   For example, if the shape of  \b grad_output is [N, T, C, HW], the shape of shifts is [N, G],
```

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -12863,6 +12863,132 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ *   C should be multiple of G, C and G must not be zero.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:12906]:
```suggestion
 *   C must be a multiple of G, and C and G cannot be zero.
```

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -12863,6 +12863,132 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
                          void *data_col);
 
+// Group:TinShift
+/*!
+ * @brief Shifts gradients from \b grad_output according to shift information in \b shifts and stores the
+ * result into \b grad_input.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * tin_shift_backward operation. For detailed information, see ::mluOpHandle_t.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:12873]: tin_shift_backward加链接~ 删除operation吧

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -12863,6 +12863,132 @@ mluOpMsDeformAttnForward(mluOpHandle_t handle,
...
+ * - The supported data types of input and output tensors are as follows:
+ *   - input tensor: half, float
+ *   - shifts tensor: int32
+ *   - output tensor: half, float
+ *
+ * @par Data Layout
+ * - The supported layout of input and output tensors must be \p MLUOP_LAYOUT_ARRAY.
+ *
+ * @par Scale Limitation
+ * - The first dimension of \b grad_output tensor and \b shifts tensor must be the same size.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:12903]:
```suggestion
 * - The first dimension of tensor \b grad_output and tensor \b shifts must be the same size.
```

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 6]
```diff
@@ -799,6 +799,18 @@ mluOpThreeNNForward
 -----------------------------
 该算子为点云 ``unknown`` 集合中的点的寻找来自 ``known``集合中的前 ``3`` 个邻近点。点云数据点的坐标为 ``(x, y, z)`` ， 通过计算平方差距离后排序，得到前3个邻近点及其在集合中的 ``index``。
 
+.. _tin_shift_backward:
+
+mluOpTinShiftBackward
+-----------------------------
+该算子为 ``mluOpTinShiftForward`` 的反向功能，给定输出数据对应的梯度、偏移量，计算输入数据对应的梯度。
```
Comment (AquamarineSu) [inline docs/bangc-docs/user_guide/9_operators/index.rst:806]: 设计文档里的几张图不错呀，要不要加几句描述，把图片加到用户手册里？

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 7]
```diff
@@ -799,6 +799,18 @@ mluOpThreeNNForward
 -----------------------------
 该算子为点云 ``unknown`` 集合中的点的寻找来自 ``known``集合中的前 ``3`` 个邻近点。点云数据点的坐标为 ``(x, y, z)`` ， 通过计算平方差距离后排序，得到前3个邻近点及其在集合中的 ``index``。
```
Comment (AquamarineSu) [inline docs/bangc-docs/user_guide/9_operators/index.rst:800]: 目测这里会报错，您帮忙加个空格？
```suggestion
该算子为点云 ``unknown`` 集合中的点的寻找来自 ``known`` 集合中的前 ``3`` 个邻近点。点云数据点的坐标为 ``(x, y, z)`` ， 通过计算平方差距离后排序，得到前3个邻近点及其在集合中的 ``index``。
```

**Review by AquamarineSu — APPROVED** [review]

**Review by DanieeelLiu — APPROVED** [review]

**Review by duzekunKTH — APPROVED** [review]

**Pull request merged**